### PR TITLE
boost: support for cross-compiling to mips64el

### PR DESCRIPTION
**Pull request by ghost**
This patch adds support for cross-compiling boost to a mips64el hostPlatform.

###### Motivation for this change

This is needed in order to bootstrap nixpkgs on a mips64el hostPlatform -- if the initial mips64el `nix` binary is to be built using nixpkgs, it needs to be cross-compiled from some other platform. 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] mips64el-linux (cross from x86_64-linux)
  - [x] aarch64-linux
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by a-m-joseph** [comment]
Previous push just rebases updates to #161159.

**Comment by a-m-joseph** [comment]
Latest push rebases against #161159 and squashes this PR into a single commit; no other changes.

File: pkgs/development/libraries/boost/generic.nix [hunk 1]
```diff
@@ -34,6 +34,11 @@ assert enableNumpy -> enablePython;
 # Boost <1.69 can't be built on linux with clang >8, because pth was removed
 assert with lib; ((stdenv.isLinux && toolset == "clang" && !(versionOlder stdenv.cc.version "8.0.0")) -> !(versionOlder version "1.69"));
 
+# boost-context lacks support for the N32 ABI on mips64.  The build
+# will succeed, but packages depending on boost-context will fail with
+# a very cryptic error message.
+assert !stdenv.hostPlatform.isMips64n32;
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/boost/generic.nix:40]: This should be done in broken instead.

File: lib/systems/parse.nix [hunk 2]
```diff
@@ -463,6 +470,7 @@ rec {
             if lib.versionAtLeast (parsed.cpu.version or "0") "6"
             then abis.gnueabihf
             else abis.gnueabi
+          else if args.cpu=="mips64el" then abis.gnuabi64  # gnuabin32 is also possible on mips64
```
Comment (SuperSandro2000) [inline lib/systems/parse.nix:473]:
```suggestion
          else if args.cpu == "mips64el" then abis.gnuabi64  # gnuabin32 is also possible on mips64
```

File: pkgs/development/libraries/boost/generic.nix (hunk 1 above)
Comment (a-m-joseph) [reply pkgs/development/libraries/boost/generic.nix]: Implemented.

**Comment by a-m-joseph** [comment]
Latest push squashes to a single commit and rebases on #161159; no other changes.

**Comment by a-m-joseph** [comment]
Rebased.  #161158 has merged so this is usable now.

**Comment by a-m-joseph** [comment]
Ping...

File: pkgs/development/libraries/boost/generic.nix [hunk 3]
```diff
@@ -114,6 +116,7 @@ stdenv.mkDerivation {
   patchFlags = [];
 
   patches = patches
+  ++ optional stdenv.hostPlatform.isMips64 ./mips64-context.patch
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/boost/generic.nix:119]: I think we should always apply this to make sure the patch works in the future and less maintenance is on you. If you take a look at the other patches they are mostly specific to boost version ranges.

Also where did you find the patch? Is there a source we can mention?

File: pkgs/development/libraries/boost/generic.nix [hunk 4]
```diff
@@ -84,15 +84,17 @@ let
     # TODO: make this unconditional
   ] ++ optionals (stdenv.hostPlatform != stdenv.buildPlatform) [
     "address-model=${toString stdenv.hostPlatform.parsed.cpu.bits}"
-    "architecture=${toString stdenv.hostPlatform.parsed.cpu.family}"
+    "architecture=${if stdenv.hostPlatform.isMips64 then "mips64" else toString stdenv.hostPlatform.parsed.cpu.family}"
+  ] ++ optionals (stdenv.hostPlatform != stdenv.buildPlatform) [
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/boost/generic.nix:88]:
```suggestion
```
That is duplicated with line 85, isn't it?

File: pkgs/development/libraries/boost/generic.nix [hunk 5]
```diff
@@ -114,6 +116,13 @@ stdenv.mkDerivation {
   patchFlags = [];
 
   patches = patches
+  ++ optional (versionOlder version "1.70") (fetchpatch {
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/boost/generic.nix:119]: I am not sure if there is any order here but maybe put it betweenline 137 and 138?

**Review by SuperSandro2000 — COMMENTED** [review]
after that I think we can merge this. Going to build it locally to see if something regresses.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).

Result of `nixpkgs-review pr 161162` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package marked as broken and skipped:
</summary>
<ul>
<li>rapcad</li>
</ul>
</details>
<details>

  <summary>6 packages failed to build and already failed to build on hydra master:</summary>
  <ul>
<li>appleseed: log was empty</li>
<li>masari: <a href="https://termbin.com/yrrl">plain log</a>  |  <a href="https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/3762f1caf8da5bb4067817e17876d7fc/raw/index.html">pretty log </a></li>
<li>osl: <a href="https://termbin.com/9713">plain log</a>  |  <a href="https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/97245b0f7c03e61bd8a5568dc8a30abd/raw/index.html">pretty log </a></li>
<li>python310Packages.appleseed: log was empty</li>
<li>python310Packages.boost169: <a href="https://termbin.com/fj4s">plain log</a>  |  <a href="https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/1b4d6346920fe68623ef560aefd8ab27/raw/index.html">pretty log </a></li>
<li>sourcetrail: log was empty</li>
  </ul>
</details>
<details>

  <summary>19 packages failed to build and are new build failures:</summary>
  <ul>
<li>amarok: log was empty</li>
<li>(amarok-kf5): log was empty</li>
<li>(python39Packages.appleseed): log was empty</li>
<li>boost155: <a href="https://termbin.com/31lq">plain log</a>  |  <a href="https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/ac19dc1276f3b14242389c644ee9ca75/raw/index.html">pretty log </a></li>
<li>boost159: <a href="https://termbin.com/7j79">plain log</a>  |  <a href="https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/b29429364d6bb18729ae06311124454a/raw/index.html">pretty log </a></li>
<li>(boost15x): log was empty</li>
<li>boost160: <a href="https://termbin.com/y5fie">plain log</a>  |  <a href="https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/e700d9302f470f36ccfe072f2108a51a/raw/index.html">pretty log </a></li>
<li>cqrlog: log was empty</li>
<li>globulation2: log was empty</li>
<li>mongodb: log was empty</li>
<li>(mongodb-3_4): log was empty</li>
<li>mongodb-3_6: log was empty</li>
<li>mongodb-4_0: <a href="https://termbin.com/20c7">plain log</a>  |  <a href="https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/42635019aa7a4f8d4cdbf9d5101baa31/raw/index.html">pretty log </a></li>
<li>mongodb-4_2: <a href="https://termbin.com/94c9">plain log</a>  |  <a href="https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/cbe0b3e84b4a2278889543e7abea67ba/raw/index.html">pretty log </a></li>
<li>mysql57: log was empty</li>
<li>percona-xtrabackup_2_4: log was empty</li>
<li>qpid-cpp: log was empty</li>
<li>spring: log was empty</li>
<li>springLobby: log was empty</li>
  </ul>
</details>
<details>
<summary>
38 packages built:
</summary>
<ul>
<li>SkypeExport</li>
<li>arc_unpacker</li>
<li>bitcoin-abc</li>
<li>bitcoin-classic</li>
<li>bitcoin-gold</li>
<li>bitcoind-abc</li>
<li>bitcoind-classic</li>
<li>bitcoind-gold</li>
<li>boost165</li>
<li>boost166</li>
<li>boost168</li>
<li>boost169 (boost16x)</li>
<li>clfft</li>
<li>cpp-netlib</li>
<li>dogecoin</li>
<li>dogecoind</li>
<li>ethminer (ethminer-cuda)</li>
<li>ethminer-free</li>
<li>foundationdb (foundationdb61)</li>
<li>freecad</li>
<li>freeorion</li>
<li>ispike</li>
<li>mirtk</li>
<li>mozart2</li>
<li>performous</li>
<li>pokerth</li>
<li>pokerth-server</li>
<li>python310Packages.foundationdb61</li>
<li>python39Packages.boost169</li>
<li>python39Packages.foundationdb61</li>
<li>starspace</li>
<li>sumokoin</li>
<li>swift-im</li>
<li>twister</li>
<li>vertcoin</li>
<li>vertcoind</li>
<li>xtreemfs</li>
<li>yafaray-core</li>
</ul>
</details>

File: pkgs/development/libraries/boost/generic.nix (hunk 5 above)
Comment (a-m-joseph) [reply pkgs/development/libraries/boost/generic.nix]: Excellent point; I've moved it.

**Comment by a-m-joseph** [comment]
> * amarok: log was empty
> 
> * (amarok-kf5): log was empty
> 
> * (python39Packages.appleseed): log was empty
> 
> * boost155: [plain log](https://termbin.com/31lq)  |  [pretty log ](https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/ac19dc1276f3b14242389c644ee9ca75/raw/index.html)
> 
> * boost159: [plain log](https://termbin.com/7j79)  |  [pretty log ](https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/b29429364d6bb18729ae06311124454a/raw/index.html)
> 
> * (boost15x): log was empty
> 
> * boost160: [plain log](https://termbin.com/y5fie)  |  [pretty log ](https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/e700d9302f470f36ccfe072f2108a51a/raw/index.html)
> 
> * cqrlog: log was empty
> 
> * globulation2: log was empty
> 
> * mongodb: log was empty
> 
> * (mongodb-3_4): log was empty
> 
> * mongodb-3_6: log was empty
> 
> * mongodb-4_0: [plain log](https://termbin.com/20c7)  |  [pretty log ](https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/42635019aa7a4f8d4cdbf9d5101baa31/raw/index.html)
> 
> * mongodb-4_2: [plain log](https://termbin.com/94c9)  |  [pretty log ](https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/cbe0b3e84b4a2278889543e7abea67ba/raw/index.html)
> 
> * mysql57: log was empty
> 
> * percona-xtrabackup_2_4: log was empty
> 
> * qpid-cpp: log was empty
> 
> * spring: log was empty
> 
> * springLobby: log was empty

Thanks, I am investigating these.

**Comment by a-m-joseph** [comment]
Resolved merge conflict with `master`.

**Comment by a-m-joseph** [comment]
Okay, latest push should fix the breakage.  The patch to backport mips64n64 support doesn't apply cleanly to anything before 1.65, but very few packages require a version of boost older than that.

**Comment by a-m-joseph** [comment]
I'm running `nixpkgs-review pr 161162` myself, it'll take a while though.

**Comment by a-m-joseph** [comment]
> I'm running `nixpkgs-review pr 161162` myself, it'll take a while though.

```
8 packages failed to build:
appleseed masari mongodb-4_0 mongodb-4_2 osl python310Packages.appleseed python310Packages.boost169 sourcetrail
```

Of these, only `mongodb-4_0` and `mongodb-4_2` were not on the "already failing before PR" list.  I tried building those at the previous commit and they failed there as well, with:

```
src/mongo/stdx/thread.h: At global scope:
src/mongo/stdx/thread.h:109:56: error: call to non-'constexpr' function 'long int sysconf(int)'
  109 |         std::max(kMongoMinSignalStackSize, std::size_t{MINSIGSTKSZ});
      |                                                        ^~~~~~~~~~~
```

So, I'm going to say that at this point this PR is not causing any additional breakage.  @SuperSandro2000, could you run your bot-tool again?  I think this is ready to merge now.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).

Result of `nixpkgs-review pr 161162` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>6 packages failed to build and already failed to build on hydra master:</summary>
  <ul>
<li>appleseed: log was empty</li>
<li>masari: <a href="https://termbin.com/3t3v">plain log</a>  |  <a href="https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/4995e45b2b5ebec616530f93eba9fd95/raw/index.html">pretty log </a></li>
<li>osl: <a href="https://termbin.com/wt7o">plain log</a>  |  <a href="https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/8f3468c1e8a9746b4372ca49da6473b1/raw/index.html">pretty log </a></li>
<li>python310Packages.appleseed: log was empty</li>
<li>python310Packages.boost169: <a href="https://termbin.com/di41">plain log</a>  |  <a href="https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/8d7a8741e85e1af08eaf7f4858f54f3a/raw/index.html">pretty log </a></li>
<li>sourcetrail: log was empty</li>
  </ul>
</details>
<details>

  <summary>3 packages failed to build and are new build failures:</summary>
  <ul>
<li>(python39Packages.appleseed): log was empty</li>
<li>mongodb-4_0: <a href="https://termbin.com/bxv1x">plain log</a>  |  <a href="https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/6eeb8a4bd5b328f5d3b5086f860cc42f/raw/index.html">pretty log </a></li>
<li>mongodb-4_2: <a href="https://termbin.com/mkcr">plain log</a>  |  <a href="https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/66b4fce7bbe4a726cdadc377f2f7a966/raw/index.html">pretty log </a></li>
  </ul>
</details>
<details>
<summary>
38 packages built:
</summary>
<ul>
<li>SkypeExport</li>
<li>arc_unpacker</li>
<li>bitcoin-abc</li>
<li>bitcoin-classic</li>
<li>bitcoin-gold</li>
<li>bitcoind-abc</li>
<li>bitcoind-classic</li>
<li>bitcoind-gold</li>
<li>boost165</li>
<li>boost166</li>
<li>boost168</li>
<li>boost169 (boost16x)</li>
<li>clfft</li>
<li>cpp-netlib</li>
<li>dogecoin</li>
<li>dogecoind</li>
<li>ethminer (ethminer-cuda)</li>
<li>ethminer-free</li>
<li>foundationdb (foundationdb61)</li>
<li>freecad</li>
<li>freeorion</li>
<li>ispike</li>
<li>mirtk</li>
<li>mozart2</li>
<li>performous</li>
<li>pokerth</li>
<li>pokerth-server</li>
<li>python310Packages.foundationdb61</li>
<li>python39Packages.boost169</li>
<li>python39Packages.foundationdb61</li>
<li>starspace</li>
<li>sumokoin</li>
<li>swift-im</li>
<li>twister</li>
<li>vertcoin</li>
<li>vertcoind</li>
<li>xtreemfs</li>
<li>yafaray-core</li>
</ul>
</details>

**Comment by a-m-joseph** [comment]
> * mongodb-4_0: [plain log](https://termbin.com/bxv1x)  |  [pretty log ](https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/6eeb8a4bd5b328f5d3b5086f860cc42f/raw/index.html)
> 
> * mongodb-4_2: [plain log](https://termbin.com/mkcr)  |  [pretty log ](https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/66b4fce7bbe4a726cdadc377f2f7a966/raw/index.html)

Yeah these also fail at the commit prior to this PR, so I think nixpkgs-review's list of "existing failures" is slightly out of date or something.

**Comment by SuperSandro2000** [comment]
> Yeah these also fail at the commit prior to this PR, so I think nixpkgs-review's list of "existing failures" is slightly out of date or something.

it uses the latest hydra eval which can be a few days behind.

**Review by hjones2199 — COMMENTED** [review]
I know its already merged, but just wanted to confirm the three failed builds seem to be from before

**Pull request merged**